http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/Http2FrameExceptionHandler.java

```diff
@@ -0,0 +1,115 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.nio.netty.internal.http2;
+
+import static software.amazon.awssdk.http.nio.netty.internal.ChannelAttributeKey.CHANNEL_POOL_RECORD;
+import static software.amazon.awssdk.http.nio.netty.internal.ChannelAttributeKey.PING_STATUS;
+import static software.amazon.awssdk.http.nio.netty.internal.ChannelAttributeKey.PING_TIMER_FUTURE;
+import static software.amazon.awssdk.http.nio.netty.internal.NettyConfiguration.HTTP2_CONNECTION_PING_TIMEOUT_SECONDS;
+import static software.amazon.awssdk.http.nio.netty.internal.utils.ChannelUtils.getAttribute;
+import static software.amazon.awssdk.http.nio.netty.internal.utils.NettyUtils.doInEventLoop;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.http2.DefaultHttp2PingFrame;
+import io.netty.handler.codec.http2.Http2PingFrame;
+import io.netty.handler.timeout.ReadTimeoutException;
+import io.netty.handler.timeout.WriteTimeoutException;
+import io.netty.util.concurrent.ScheduledFuture;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Logger;
+
+@ChannelHandler.Sharable
+@SdkInternalApi
+public final class Http2FrameExceptionHandler extends ChannelInboundHandlerAdapter {
+    private static final Http2FrameExceptionHandler INSTANCE = new Http2FrameExceptionHandler();
+    private static final Http2PingFrame DEFAULT_PING_FRAME = new DefaultHttp2PingFrame(0);
+    private static final Logger log = Logger.loggerFor(Http2FrameExceptionHandler.class);
+
+    private Http2FrameExceptionHandler() {
+    }
+
+    public static Http2FrameExceptionHandler create() {
+        return INSTANCE;
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
+        if ((cause instanceof ReadTimeoutException || cause instanceof WriteTimeoutException) && ctx.channel().parent() != null) {
+            Channel parent = ctx.channel().parent();
+            doInEventLoop(parent.eventLoop(), () -> healthCheckIfNotAlready(parent));
+        }
+
+        ctx.fireExceptionCaught(cause);
+    }
+
+    private void healthCheckIfNotAlready(Channel parent) {
+
+        // No need to do another health check if there is a PING in flight
+        if (PingStatus.IN_FLIGHT.equals(parent.attr(PING_STATUS).get())) {
```

Comment:
Can we consolidate PING_STATUS and PING_TIMER_FUTURE?

If we get a read timeout:
1. We schedule a "shut down after 5 seconds" task.
2. We send a ping to the service

If we get a ping response:
1. We cancel the "shut down after 5 seconds" task.